packages/block-editor/src/store/actions.js

```diff
@@ -256,16 +296,51 @@ export const moveBlocksUp = createOnMove( 'MOVE_BLOCKS_UP' );
  * @param  {?string} toRootClientId   Root client ID destination.
  * @param  {number}  index            The index to move the block into.
  *
- * @return {Object} Action object.
+ * @yields {Object} Action object.
  */
-export function moveBlockToPosition( clientId, fromRootClientId, toRootClientId, index ) {
-	return {
+export function* moveBlockToPosition( clientId, fromRootClientId, toRootClientId, index ) {
+	const templateLock = yield select(
+		'core/block-editor',
+		'getTemplateLock',
+		fromRootClientId
+	);
+
+	// If locking is equal to all on the original clientId (fromRootClientId),
+	// it is not possible to move the block to any other position.
+	if ( templateLock === 'all' ) {
+		return;
+	}
+
+	const action = {
 		type: 'MOVE_BLOCK_TO_POSITION',
 		fromRootClientId,
 		toRootClientId,
 		clientId,
 		index,
 	};
+	// If moving inside the same root block the move is always possible.
```

Comment:
Hi @aduth although this could be seen as optimization as the majority of cases will return here, this is also required.
The case this covers is locking=insert, in that case, we can not insert blocks but we can move the blocks, canInsertBlockType will return false but the move is in fact possible. Without this condition when locking = insert we would not be able to move the blocks as expected.